### PR TITLE
Fix warnings for undefined array keys

### PR DESCRIPTION
**Pull request by rabauss**
This PR fixes the following warnings with PHP 8 while cache warmup:
```php
PHP Warning:  Undefined array key "GATEWAY" in /app/vendor/terminal42/notification_center/config/config.php on line 95
PHP Warning:  Undefined array key "NOTIFICATION_TYPE" in /app/vendor/terminal42/notification_center/config/config.php on line 108
```

I fixed the other warning for missing translation key with `\Contao\System::loadLanguageFile('tl_nc_queue');`. First I was not sure if it's a bug from the core:
```php
PHP Warning:  Undefined array key "re-queueConfirmation" in /app/var/cache/prod/contao/dca/tl_nc_queue.php on line 19
```

### CODE REVIEW

File: config/config.php [hunk 1]
```diff
@@ -91,6 +91,9 @@
 /**
  * Notification Center Gateways
  */
+if (!isset($GLOBALS['NOTIFICATION_CENTER']['GATEWAY'])) {
+    $GLOBALS['NOTIFICATION_CENTER']['GATEWAY'] = [];
+}
 $GLOBALS['NOTIFICATION_CENTER']['GATEWAY'] = array_merge(
     (array) $GLOBALS['NOTIFICATION_CENTER']['GATEWAY'],
```
Comment (aschempp) [inline config/config.php:98]: To be consistent with Contao, shouldn't we fix it like that?

```suggestion
    (array) ($GLOBALS['NOTIFICATION_CENTER']['GATEWAY'] ?? null),
```

Comment (rabauss) [reply config/config.php:98]: Yes, but in composer.json there is still PHP 5.4 as allowed requirement!

File: dca/tl_nc_queue.php [hunk 2]
```diff
@@ -8,6 +8,8 @@
  * @license    LGPL
  */
 
+\Contao\System::loadLanguageFile('tl_nc_queue');
```
Comment (rabauss) [inline dca/tl_nc_queue.php:11]: Is this the right solution for the missing translation for `re-queueConfirmation` while `cache:warmup`?

**Comment by rabauss** [comment]
Any suggestions to merge this PR?

File: config/config.php (hunk 1 above)
Comment (aschempp) [reply config/config.php:98]: 🤦 😂

File: dca/tl_nc_queue.php (hunk 2 above)
Comment (aschempp) [reply dca/tl_nc_queue.php:11]: That does not look right to me, that's the reason for this?

Comment (rabauss) [reply dca/tl_nc_queue.php:11]: while `cache:warmup` there is a warning for a missing translation array key `re-queueConfirmation`

Comment (rabauss) [reply dca/tl_nc_queue.php:11]: Are there any better solutions?

Comment (m-vo) [reply dca/tl_nc_queue.php:11]: In the core we added a null check to the usage. Like this:

```php
if(!confirm(\'' . $GLOBALS['TL_LANG']['tl_nc_queue']['re-queueConfirmation'] ?? null . '\')) […]
```

**Comment by christianromeni** [comment]
I just came across this when trying out php8 ... 
image

Thx for working ahead! ;)

File: dca/tl_nc_queue.php (hunk 2 above)
Comment (fritzmg) [reply dca/tl_nc_queue.php:11]: Possibly related to https://github.com/contao/contao/pull/2536 ?

The line should not be necessary.

**Review by fritzmg — CHANGES_REQUESTED** [review]

File: dca/tl_nc_queue.php (hunk 2 above)
Comment (rabauss) [reply dca/tl_nc_queue.php:11]: But if I do not load the language file I get the following error while `cache:warmup`: `PHP Warning:  Undefined array key "re-queueConfirmation"`

Comment (fritzmg) [reply dca/tl_nc_queue.php:11]: Hm, I could not reproduce. I installed current `dev-main` with PHP `8.0.9` and there were no errors during `cache:warmup`.

File: library/NotificationCenter/AutoSuggester.php [hunk 3]
```diff
@@ -42,7 +42,7 @@ public static function load($dc)
         static::$strType = static::$objNotification->type;
 
         foreach ($GLOBALS['TL_DCA'][static::$strTable]['fields'] as $field => $arrConfig) {
-            if ('nc_tokens' === $arrConfig['eval']['rgxp']) {
+            if (isset($arrConfig['eval']['rgxp']) && 'nc_tokens' === $arrConfig['eval']['rgxp']) {
```
Comment (fritzmg) [inline library/NotificationCenter/AutoSuggester.php:45]:
```suggestion
            if ('nc_tokens' === $arrConfig['eval']['rgxp'] ?? null) {
```

File: library/NotificationCenter/AutoSuggester.php [hunk 4]
```diff
@@ -69,7 +69,11 @@ public function init(\DataContainer $dc)
         $arrParsedTokens = array();
 
         foreach ($arrTokens as $strToken) {
-            $content = $GLOBALS['TL_LANG']['NOTIFICATION_CENTER_TOKEN'][static::$strType][$strToken] ?: '';
+            $content = '';
+
+            if (isset($GLOBALS['TL_LANG']['NOTIFICATION_CENTER_TOKEN'][static::$strType][$strToken])) {
+                $content = $GLOBALS['TL_LANG']['NOTIFICATION_CENTER_TOKEN'][static::$strType][$strToken] ?: '';
+            }
```
Comment (fritzmg) [inline library/NotificationCenter/AutoSuggester.php:76]:
```suggestion
            $content = $GLOBALS['TL_LANG']['NOTIFICATION_CENTER_TOKEN'][static::$strType][$strToken] ?? '';
```
Should be sufficient, probably.

**Review by fritzmg — CHANGES_REQUESTED** [review]

File: library/NotificationCenter/AutoSuggester.php (hunk 3 above)
Comment (rabauss) [reply library/NotificationCenter/AutoSuggester.php:45]: Unfortunately, there is still support for `php ^5.4` in `composer.json` - this is not compatible - shall we update the `php` requirement?

File: library/NotificationCenter/AutoSuggester.php (hunk 4 above)
Comment (rabauss) [reply library/NotificationCenter/AutoSuggester.php:76]: Unfortunately, there is still support for `php ^5.4` in `composer.json` - this is not compatible - shall we update the `php` requirement?

Comment (fritzmg) [reply library/NotificationCenter/AutoSuggester.php:76]: Oh right, I did not notice. Yes, increase the version requirement. Supporting PHP 5 is not necessary at all anymore in my opinion.

File: dca/tl_nc_queue.php (hunk 2 above)
Comment (fritzmg) [reply dca/tl_nc_queue.php:11]: @rabauss can you try to reproduce this again and post the complete stack trace?

**Review by fritzmg — CHANGES_REQUESTED** [review]

File: dca/tl_nc_queue.php (hunk 2 above)
Comment (rabauss) [reply dca/tl_nc_queue.php:11]: For me the Language file of `tl_nc_queue` is really not loaded while `cache:warmup` - here is the callstack:

```
tl_nc_queue.php:17, include()
DcaLoader.php:93, Contao\DcaLoader->loadDcaFiles()
DcaLoader.php:70, Contao\DcaLoader->load()
Controller.php:1360, Contao\Controller::loadDataContainer()
DcaExtractor.php:380, Contao\DcaExtractor->createExtract()
DcaExtractor.php:128, Contao\DcaExtractor->__construct()
DcaExtractor.php:150, Contao\DcaExtractor::getInstance()
ContaoCacheWarmer.php:197, Contao\CoreBundle\Cache\ContaoCacheWarmer->generateDcaExtracts()
ContaoCacheWarmer.php:94, Contao\CoreBundle\Cache\ContaoCacheWarmer->warmUp()
CacheWarmerAggregate.php:98, Symfony\Component\HttpKernel\CacheWarmer\CacheWarmerAggregate->warmUp()
CacheWarmupCommand.php:81, Symfony\Bundle\FrameworkBundle\Command\CacheWarmupCommand->execute()
Command.php:256, Symfony\Bundle\FrameworkBundle\Command\CacheWarmupCommand->run()
Application.php:989, Contao\ManagerBundle\Console\ContaoApplication->doRunCommand()
Application.php:96, Contao\ManagerBundle\Console\ContaoApplication->doRunCommand()
Application.php:290, Contao\ManagerBundle\Console\ContaoApplication->doRun()
Application.php:82, Contao\ManagerBundle\Console\ContaoApplication->doRun()
Application.php:166, Contao\ManagerBundle\Console\ContaoApplication->run()
contao-console:37, {main}()
```

Comment (fritzmg) [reply dca/tl_nc_queue.php:11]: Is that the stack trace for the warning?

Comment (rabauss) [reply dca/tl_nc_queue.php:11]: Oh crazy, now I had some trouble with my docker containers and I had to rebuild my project with php 8.0.10 the warning does not occur anymore without the line `\Contao\System::loadLanguageFile('tl_nc_queue')`

Comment (fritzmg) [reply dca/tl_nc_queue.php:11]: I still cannot reproduce with PHP 8 and Contao 4.9. You should never have to load the language file of the same DCA yourself. You only need to load language files, if your DCA accesses translations of different DCAs/language domains.

Comment (rabauss) [reply dca/tl_nc_queue.php:11]: I can reproduce it only with an existing database!

Comment (fritzmg) [reply dca/tl_nc_queue.php:11]: Well yes I am also testing with an existing database :).

Comment (rabauss) [reply dca/tl_nc_queue.php]: Thanks @fritzmg  for the support - for me the js-warning for `re-queueConfirmation` is also working with the `??` check!

File: config/config.php [hunk 5]
```diff
@@ -104,6 +107,9 @@
 /**
  * Notification Center Notification Types
  */
+if (!isset($GLOBALS['NOTIFICATION_CENTER']['NOTIFICATION_TYPE'])) {
+    $GLOBALS['NOTIFICATION_CENTER']['NOTIFICATION_TYPE'] = [];
+}
 $GLOBALS['NOTIFICATION_CENTER']['NOTIFICATION_TYPE'] = array_merge_recursive(
     (array) $GLOBALS['NOTIFICATION_CENTER']['NOTIFICATION_TYPE'],
```
Comment (fritzmg) [inline config/config.php:114]: Now that we dropped PHP 5 support you can change it to

```php
$GLOBALS['NOTIFICATION_CENTER']['NOTIFICATION_TYPE'] = array_merge_recursive(
    $GLOBALS['NOTIFICATION_CENTER']['NOTIFICATION_TYPE'] ?? [],
    array(
```

**Review by fritzmg — CHANGES_REQUESTED** [review]

**Review by rabauss — COMMENTED** [review]
Are we allowed to remove the deprecated class `String`?

**Comment by fritzmg** [comment]
> Are we allowed to remove the deprecated class `String`?

I would leave it to keep this PR as concise as possible. It is already deprecated anyway and thus can be removed in the next major version. The exception can still be useful to those who are updating from PHP 5 to 7 and have not migrated these things yet.

**Review by fritzmg — APPROVED** [review]

**Comment by fritzmg** [comment]
@aschempp @qzminski @Toflar any objections to this, particularly the minimum PHP version increase? Otherwise I'd merge and release this.

**Comment by Toflar** [comment]
Nah that's totally fine of course :)

**Pull request merged**